File: Python3.x/79-Word_Search.py
# dfs
class Solution:
    def exist(self, board, word):
        if not board: return False
        for i in range(len(board)):
            for j in range(len(board[0])):
                if self.dfs(board, word, i, j): return True  # 起点
        return False

    def dfs(self, board, word, i, j):  # 从i，j为开始检查是否能找到单词
        if len(word) == 0: return True  # 所有单词检查完毕
        if i < 0 or i >= len(board) or j < 0 or j >= len(board[0]) or word[0] != board[i][j]:
            return False
        tmp = board[i][j]  # 第一个字符已经被检查
        board[i][j] = '#'  # 标记已经检查过的的字符，避免重复检查
        res = self.dfs(board, word[1:], i + 1, j) or self.dfs(board, word[1:], i - 1, j) or \
            self.dfs(board, word[1:], i, j + 1) or self.dfs(board, word[1:], i, j - 1)  # 检查是否能沿着一个方向匹配
        board[i][j] = tmp
        return res

File: Algorithms/C++/79-Word_Search.cpp
// Runtime: 16 ms, faster than 98.95% of C++ online submissions for Word Search.
// Memory Usage: 10 MB, less than 100.00% of C++ online submissions for Word Search.
class Solution {
public:
    bool exist(vector<vector<char>>& board, string word) {
        len_row = board.size();
        len_col = board[0].size();
        for (int row = 0; row < len_row; ++row)
            for (int col = 0; col < len_col; ++col)
                if (is_found(board, word.c_str(), row, col))
                    return true;
        return false;
    }
private:
    int len_row, len_col;
    bool is_found(vector<vector<char>>& board, const char* w, int row, int col) {
        if (row < 0 || col < 0 || row >= len_row || col >= len_col || board[row][col] == '\0' || board[row][col] != *w)
            return false;
        if (*(w + 1) == '\0')
            return true;
        char temp = board[row][col];
        board[row][col] = '\0';
        if (is_found(board, w + 1, row - 1, col) || is_found(board, w + 1, row + 1, col) ||
            is_found(board, w + 1, row, col - 1) || is_found(board, w + 1, row, col + 1))
            return true;
        board[row][col] = temp;
        return false;
    }
};
